perun-core/src/main/java/cz/metacentrum/perun/core/provisioning/GenDataProviderImpl.java

```diff
@@ -0,0 +1,358 @@
+
+		return getAndStoreHash(hash, group, groupAttrs);
+	}
+
+	private List<String> getGroupResourceAttributesHashes(Group group, Resource resource) {
+		if (!resource.equals(lastLoadedResource)) {
+			throw new IllegalStateException("The last loaded resource is different than the required one. Required: " +
+					resource + ", Last loaded: " + lastLoadedResource);
+		}
+
+		String hash = hasher.hashGroupResource(group, resource);
+
+		return getAndStoreHash(hash, group, groupResourceAttrs);
+	}
+
+	private List<String> getUserAttributesHashes(User user) {
+		String hash = hasher.hashUser(user);
+
+		return getAndStoreHash(hash, user, userAttrs);
+	}
+
+	private List<String> getUserFacilityAttributesHashes(User user, Facility facility) {
+		String hash = hasher.hashUserFacility(user, facility);
+
+		return getAndStoreHash(hash, user, userFacilityAttrs);
+	}
+
+	/**
+	 * Get a list of attributes for given hash, and loads appropriate entity attributes from given map
+	 * into the map of all processed attributes.
+	 *
+	 * If the map doesn't contain attributes for the given entity, or the list is empty,
+	 * this method returns an empty list. Otherwise, it returns a List with the given hash.
+	 *
+	 * @param hash entity hash
+	 * @param entity the entity which the hash belongs
+	 * @param map map of attributes for the entity
+	 * @param <T> the type of the entity User, Member, ...
+	 * @return List with the hash or empty list if the map doesn't contain any attributes for the given entity
+	 */
+	private <T> List<String> getAndStoreHash(String hash, T entity, Map<T, List<Attribute>> map) {
+		if (!attributesByHash.containsKey(hash)) {
+			if (!map.containsKey(entity)) {
+				return emptyList();
+			}
+			attributesByHash.put(hash, map.get(entity));
+		}
+		return attributesByHash.get(hash).isEmpty() ? emptyList() : singletonList(hash);
+	}
+
+	private void loadMemberSpecificAttributes(List<Member> members) {
+		memberAttrs.putAll(
+				sess.getPerunBl().getAttributesManagerBl().getRequiredAttributes(sess, null, service, members)
+		);
+
+		List<Integer> userIds = members.stream()
+				.map(Member::getUserId)
+				.collect(toList());
+
+		List<User> users = sess.getPerunBl().getUsersManagerBl().getUsersByIds(sess, userIds);
+
+		Map<Integer, User> usersById = users.stream()
+				.collect(toMap(User::getId, Function.identity()));
+		loadedUsersById.putAll(usersById);
+
+		loadUserSpecificAttributes(users);
+	}
+
+	private void loadUserSpecificAttributes(List<User> users) {
+		userAttrs.putAll(
+				sess.getPerunBl().getAttributesManagerBl().getRequiredAttributes(sess, service, users)
+		);
+
+		userFacilityAttrs.putAll(
+				sess.getPerunBl().getAttributesManagerBl().getRequiredAttributes(sess, service, facility, users)
+		);
+	}
+
+	private void loadVoSpecificAttributes(Resource resource) {
+		// create fake vo for map, because of optimization
+		Vo vo = new Vo();
+		vo.setId(resource.getVoId());
+		if (!voAttrs.containsKey(vo)) {
```

Comment:
There is probably missing assigning of actual processed resource object to variable lastLoadedResource.

Reply:
I was missing because this was a private method and at the place where it was used, the last loaded resource was set. But I have added it here as well just to make sure.